diff -ur ./otp_src_R13B04.orig/erts/emulator/beam/dist.c ./otp_src_R13B04/erts/emulator/beam/dist.c
--- ./otp_src_R13B04.orig/erts/emulator/beam/dist.c	2009-11-20 05:29:24.000000000 -0800
+++ ./otp_src_R13B04/erts/emulator/beam/dist.c	2010-02-14 15:06:32.000000000 -0800
@@ -97,6 +97,9 @@
 #define PASS_THROUGH 'p'        /* This code should go */
 
 int erts_is_alive; /* System must be blocked on change */
+#define ERTS_DE_BUSY_LIMIT (128*1024)
+int erts_de_busy_limit = ERTS_DE_BUSY_LIMIT;
+
 
 /* distribution trap functions */
 Export* dsend2_trap = NULL;
@@ -1407,8 +1410,6 @@
     return -1;
 }
 
-#define ERTS_DE_BUSY_LIMIT (128*1024)
-
 static int
 dsig_send(ErtsDSigData *dsdp, Eterm ctl, Eterm msg, int force_busy)
 {
diff -ur ./otp_src_R13B04.orig/erts/emulator/beam/erl_init.c ./otp_src_R13B04/erts/emulator/beam/erl_init.c
--- ./otp_src_R13B04.orig/erts/emulator/beam/erl_init.c	2009-09-18 07:07:46.000000000 -0700
+++ ./otp_src_R13B04/erts/emulator/beam/erl_init.c	2010-02-14 16:08:33.000000000 -0800
@@ -788,6 +788,7 @@
     char envbuf[21]; /* enough for any 64-bit integer */
     size_t envbufsz;
     int async_max_threads = erts_async_max_threads;
+    extern int erts_de_busy_limit;
 
     early_init(&argc, argv);
 
@@ -807,6 +808,9 @@
     if (erts_sys_getenv("ERL_THREAD_POOL_SIZE", envbuf, &envbufsz) == 0) {
 	async_max_threads = atoi(envbuf);
     }
+    if (erts_sys_getenv("ERTS_DE_BUSY_LIMIT", envbuf, &envbufsz) == 0) {
+	erts_de_busy_limit = atoi(envbuf);
+    } 
     
 
 #ifdef DEBUG
